torch/csrc/autograd/engine.cpp

```diff
@@ -201,8 +216,22 @@ auto ReadyQueue::pop() -> FunctionTask {
 
 Engine::Engine() = default;
 
-// This Engine's ReadyQueues and their corresponding threads are leaked here
-Engine::~Engine() = default;
+// Send shutdown tasks to all ReadyQueues if no backward tasks are running
+// Even though readyQueue should be empty, shutdown tasks have the highest
+// priority
+Engine::~Engine() {
+  bool noBackward = true;
+  for (auto& queue: ready_queues_) {
+    std::lock_guard<std::mutex> lock(queue->mutex_);
+    noBackward =  noBackward && queue->heap_.empty();
+  }
+  if (noBackward) {
+    for (auto& queue : ready_queues_) {
+     queue->pushShutdownTask();
+    }
+  }
+  // Othewise threads are leaked
```

Comment:
What was the issue with leaking threads? This is only a best-effort measure anyway, right?